Add from_receiver method to FormulaEngine

matthias-wende-frequenz opened the pull request.

Marenz commented:
CI failing

llucax commented on src/frequenz/sdk/timeseries/_formula_engine/_formula_engine.py:
```diff
@@ -374,6 +374,24 @@ def new_receiver(
 
+    @staticmethod
+    def from_receiver(receiver: Receiver, name: str) -> FormulaEngine:
+        """
+        Create a formula engine from a receiver.
```
```suggestion
        """Create a formula engine from a receiver.
```

llucax commented on src/frequenz/sdk/timeseries/_formula_engine/_formula_engine.py:
```diff
@@ -374,6 +374,24 @@ def new_receiver(
+            A formula engine that streams the `Sample`s from the receiver.
+        """
+        builder = FormulaBuilder(name)
+        builder.push_metric(name, receiver, nones_are_zeros=False)
```
Shouldn't `nones_are_zeros` be an option to this method?

matthias-wende-frequenz replied:
You are right!

llucax reviewed (commented):
It would be nice to have a test for this, and to write what motivated this in the PR description or commit message.

matthias-wende-frequenz commented:
updated and rebased onto v0.x.x. ~Only tests are still missing.~

matthias-wende-frequenz committed:
Add `from_receiver` method to FormulaEngine

When composing two formula engines it's not possible to access the
individual elements. In such a case one can first alter the
individual streams before composing. But altering is only supported
for channel receivers while composing only works on formula engines.

This method can be used to construct a formula engine from a
receiver even after the receivers samples have been manipulated.

Signed-off-by: Matthias Wende <matthias.wende@frequenz.com>

matthias-wende-frequenz committed:
Add test for `from_receiver` method

Signed-off-by: Matthias Wende <matthias.wende@frequenz.com>

shsms reviewed (approved):
A longer term alternative for this would be to bring back specialized formula receivers,  and allow for them to be composed with engines directly, to reduce boiler plate.

llucax reviewed (commented):
Release notes! @matthias-wende-frequenz
